go/tools/gazelle/packages/fileinfo.go

```diff
@@ -0,0 +1,538 @@
+limitations under the License.
+*/
+
+package packages
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+// fileInfo holds information used to decide how to build a file. This
+// information comes from the file's name, from package and import declarations
+// (in .go files), and from +build and cgo comments.
+type fileInfo struct {
+	path, dir, name, ext string
+
+	// packageName is the Go package name of a .go file, without the
+	// "_test" suffix if it was present. It is empty for non-Go files.
+	packageName string
+
+	// category is the type of file, based on extension.
+	category extCategory
+
+	// isTest is true if the file stem (the part before the extension)
+	// ends with "_test.go". This is never true for non-Go files.
+	isTest bool
+
+	// isXTest is true for test Go files whose declared package name ends
+	// with "_test".
+	isXTest bool
+
+	// imports is a list of packages imported by a file. It does not include
+	// "C" or anything from the standard library.
+	imports []string
+
+	// isCgo is true for .go files that import "C".
+	isCgo bool
+
+	// goos and goarch contain the OS and architecture suffixes in the filename,
+	// if they were present.
+	goos, goarch string
+
+	// tags is a list of build tag lines. Each entry is the trimmed text of
+	// a line after a "+build" prefix.
+	tags []string
+
+	// copts and clinkopts contain flags that are part of CFLAGS, CPPFLAGS,
+	// CXXFLAGS, and LDFLAGS directives in cgo comments.
+	copts, clinkopts []taggedOpts
+}
+
+// taggedOpts a list of compile or link options which should only be applied
+// if the given set of build tags are satisfied.
+type taggedOpts struct {
+	tags string
+	opts []string
+}
+
+// extCategory indicates how a file should be treated, based on extention.
+type extCategory int
+
+const (
+	// ignoredExt is applied to files which are not part of a build.
+	ignoredExt extCategory = iota
+
+	// unsupportedExt is applied to files that we don't support but would be
+	// built with "go build".
+	unsupportedExt
+
+	// goExt is applied to .go files.
+	goExt
+
+	// cExt is applied to C and C++ files.
+	cExt
+
+	// hExt is applied to header files. If cgo code is present, these may be
+	// C or C++ headers. If not, they are treated as Go assembly headers.
+	hExt
+
+	// sExt is applied to Go assembly files, ending with .s.
+	sExt
+
+	// csExt is applied to other assembly files, ending with .S. These are built
+	// with the C compiler if cgo code is present.
+	csExt
+)
+
+// fileNameInfo returns information that can be inferred from the name of
+// a file. It does not read data from the file.
+func fileNameInfo(dir, name string) fileInfo {
+	ext := path.Ext(name)
+
+	// Determine test, goos, and goarch. This is intended to match the logic
+	// in goodOSArchFile in go/build.
+	var isTest bool
+	var goos, goarch string
+	i := strings.Index(name, "_")
```

Comment:
Done. This was copied from go/build originally, but I agree this is more readable.